Intersect.Client/Entities/Critter.cs

```diff
@@ -75,78 +75,98 @@ public override bool Update()
-                            Dir = 1;
-                            OffsetY = -Options.TileHeight;
-                            OffsetX = 0;
-                        }
+            var deltaX = 0;
+            var deltaY = 0;
 
-                        break;
-                    case 2: // Left
-                        if (IsTileBlocked(X - 1, Y, Z, MapId, ref blockedBy, true, true, mAttribute.IgnoreNpcAvoids) == -1 && X > 0 && (!mAttribute.BlockPlayers || !PlayerOnTile(MapId, X - 1, Y)))
-                        {
-                            tmpX--;
-                            IsMoving = true;
-                            Dir = 2;
-                            OffsetY = 0;
-                            OffsetX = Options.TileWidth;
-                        }
+            switch (MoveDir)
+            {
+                case Direction.Up:
+                    deltaX = 0;
+                    deltaY = -1;
+                    break;
+                case Direction.Down:
+                    deltaX = 0;
+                    deltaY = 1;
+                    break;
+                case Direction.Left:
+                    deltaX = -1;
+                    deltaY = 0;
+                    break;
+                case Direction.Right:
+                    deltaX = 1;
+                    deltaY = 0;
+                    break;
+                case Direction.UpLeft:
+                    deltaX = -1;
+                    deltaY = -1;
+                    break;
+                case Direction.UpRight:
+                    deltaX = 1;
+                    deltaY = -1;
+                    break;
+                case Direction.DownLeft:
+                    deltaX = -1;
+                    deltaY = 1;
+                    break;
+                case Direction.DownRight:
+                    deltaX = 1;
+                    deltaY = 1;
+                    break;
+            }
 
-                        break;
-                    case 3: // Right
-                        if (IsTileBlocked(X + 1, Y, Z, MapId, ref blockedBy, true, true, mAttribute.IgnoreNpcAvoids) == -1 && X < Options.MapWidth - 1 && (!mAttribute.BlockPlayers || !PlayerOnTile(MapId, X + 1, Y)))
-                        {
-                            //If BlockPlayers then make sure there is no player here
-                            tmpX++;
-                            IsMoving = true;
-                            Dir = 3;
-                            OffsetY = 0;
-                            OffsetX = -Options.TileWidth;
-                        }
+            var isBlocked = IsTileBlocked(tmpX + deltaX, tmpY + deltaY, Z, MapId, ref blockedBy, true, true,
+                mAttribute.IgnoreNpcAvoids) == -1;
+            var playerOnTile = PlayerOnTile(MapId, tmpX + deltaX, tmpY + deltaY);
 
-                        break;
-                }
+            if (isBlocked && tmpX + deltaX >= 0 && tmpX + deltaX < Options.MapWidth && tmpY + deltaY >= 0 &&
```

Comment:
I _think_ you might need to also check to see if at least one of deltaX and deltaY are non-zero because of the default case doing nothing.

1. Add `var skip = deltaX == 0 && deltaY == 0;` _before_ `var isBlocked = ...`
2. Move the `var isBlocked`, `var playerOnTile`, and the `if (isBlocked ...` into an `if (!skip) { ... }`. This is to avoid calling `IsTileBlocked()` and `PlayerOnTile()` if their values are going to be skipped.

Reply:
```csharp
            var skip = deltaX == 0 && deltaY == 0;

            if (!skip)
            {
                var isBlocked = IsTileBlocked(tmpX + deltaX, tmpY + deltaY, Z, MapId, ref blockedBy, true, true,
                    mAttribute.IgnoreNpcAvoids) == -1;
                var playerOnTile = PlayerOnTile(MapId, tmpX + deltaX, tmpY + deltaY);

                if (isBlocked && tmpX + deltaX >= 0 && tmpX + deltaX < Options.MapWidth && tmpY + deltaY >= 0 &&
                    tmpY + deltaY < Options.MapHeight && (!mAttribute.BlockPlayers || !playerOnTile))
                {
                    tmpX += (sbyte)deltaX;
                    tmpY += (sbyte)deltaY;
                    IsMoving = true;
                    Dir = MoveDir;

                    if (deltaX == 0)
                    {
                        OffsetX = 0;
                    }
                    else
                    {
                        OffsetX = deltaX > 0 ? -Options.TileWidth : Options.TileWidth;
                    }

                    if (deltaY == 0)
                    {
                        OffsetY = 0;
                    }
                    else
                    {
                        OffsetY = deltaY > 0 ? -Options.TileHeight : Options.TileHeight;
                    }
                }
            }

            if (IsMoving)
            {
                X = (byte)tmpX;
                Y = (byte)tmpY;
                MoveTimer = Timing.Global.Milliseconds + (long)GetMovementTime();
            }
            else if (MoveDir != Dir)
            {
                Dir = MoveDir;
            }
```